revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,232 @@
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			certResults[i] = certOCSPStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return certResults
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert, issuer *x509.Certificate, opts Options) []error {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return []error{NoOCSPServerError{}}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		status, err := checkStatusFromServer(cert, issuer, server, opts)
+		if err != nil {
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		// A valid response has been received from an OCSP server
+		// Result should be based on only this response, not any errors from
+		// other servers
+		switch status {
+		case ocsp.Revoked:
+			return []error{RevokedError{}}
+		case ocsp.Good:
+			return []error{nil}
+		default:
+			return []error{UnknownStatusError{}}
+		}
+
+	}
+	return serverErrs
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) (ocsp.ResponseStatus, error) {
+	// Check valid server
+	if !strings.HasPrefix(strings.ToLower(server), "http://") {
+		// This function is only able to check servers that are accessible via HTTP
+		return ocsp.Unknown, CheckOCSPError{Err: errors.New("OCSPServer must be accessible over HTTP")}
+	}
+
+	// Create OCSP Request
+	resp, err := ocspRequest(cert, issuer, server, opts)
+	if err != nil {
+		// If there is a server error, attempt all servers before determining what to return
+		// to the user
+		if errors.Is(err, TimeoutError{}) {
+			return ocsp.Unknown, err
+		}
+		return ocsp.Unknown, CheckOCSPError{Err: err}
+	}
+
+	// Validate OCSP response isn't expired
+	if time.Now().After(resp.NextUpdate) {
+		return ocsp.Unknown, CheckOCSPError{Err: errors.New("expired OCSP response")}
+	}
+
+	// Check for presence of pkix-ocsp-no-check extension in response
+	foundNoCheck := false
+	for _, extension := range resp.Extensions {
+		if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+			foundNoCheck = true
+			break
+		}
+	}
+	if !foundNoCheck {
+		// This will be ignored until CRL is implemented
+		// If it isn't found, CRL should be used to verify the OCSP response
+		// TODO: add CRL support
+		// https://github.com/notaryproject/notation-core-go/issues/125
+		return ocsp.Unknown, PKIXNoCheckError{}
+	}
```

Comment:
This would invert the logic. The correct response should be as follows:
 - If the extension is present, then the response is okay.
 - If the extension is missing, we need to return the error.
   - These instances can be validated by checking either CRL or via another method specified via AIA. Our spec says that we will check these via CRL in the future (meaning we should not support other arbitrary mechanisms of revocation checking), but at the moment we only support OCSP checking

Reference: https://www.ietf.org/rfc/rfc2560.txt?number=2560 4.2.2.2.1